dev/cuda/gelu_backward.cu

```diff
@@ -0,0 +1,209 @@
+// CPU code reference
+
+#define GELU_SCALING_FACTOR sqrtf(2.0f / M_PI)
+
+void gelu_backward_cpu(float* dinp, const float* inp, const float* dout, const int N) {
+    for (int i = 0; i < N; i++) {
+        float x = inp[i];
+        float cube = 0.044715f * x * x * x;
+        float tanh_arg = GELU_SCALING_FACTOR * (x + cube);
+        float tanh_out = tanhf(tanh_arg);
+        float coshf_out = coshf(tanh_arg);
+        float sech_out = 1.0f / (coshf_out * coshf_out);
+        float local_grad = 0.5f * (1.0f + tanh_out) + x * 0.5f * sech_out * GELU_SCALING_FACTOR * (1.0f + 3.0f * 0.044715f * x * x);
+        dinp[i] = (floatX)(local_grad * (float)dout[i]);
+    }
+}
+
+// ----------------------------------------------------------------------------
+// GPU kernels
+
+// elementwise ops are nice and ez
+__global__ void gelu_backward1(floatX* dinp, const floatX* inp, const floatX* dout, int N) {
+    int i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i < N) {
+        float x = (float)inp[i];
+        float cube = 0.044715f * x * x * x;
+        float tanh_arg = GELU_SCALING_FACTOR * (x + cube);
+        float tanh_out = tanhf(tanh_arg);
+        float coshf_out = coshf(tanh_arg);
+        float sech_out = 1.0f / (coshf_out * coshf_out);
+        float local_grad = 0.5f * (1.0f + tanh_out) + x * 0.5f * sech_out * GELU_SCALING_FACTOR * (1.0f + 3.0f * 0.044715f * x * x);
+        dinp[i] = (floatX)(local_grad * (float)dout[i]);
+    }
+}
+
+__global__ void gelu_backward2(floatX* dinp, const floatX* inp, const floatX* dout, const int N) {
+    int i = (blockIdx.x * blockDim.x + threadIdx.x) * x128::size;
+    if (i < N) {
+        x128 packed_dinp;
+        x128 packed_inp = load128cs(inp + i);
+        x128 packed_dout = load128cs(dout + i);
+        for (int k = 0; k < packed_inp.size; ++k) {
+            float x = (float)packed_inp[k];
+            float cube = 0.044715f * x * x * x;
+            float tanh_arg = GELU_SCALING_FACTOR * (x + cube);
+            float tanh_out = tanhf(tanh_arg);
+            float coshf_out = coshf(tanh_arg);
+            float sech_out = 1.0f / (coshf_out * coshf_out);
+            float local_grad = 0.5f * (1.0f + tanh_out) + x * 0.5f * sech_out * GELU_SCALING_FACTOR * (1.0f + 3.0f * 0.044715f * x * x);
+            packed_dinp[k] = (floatX)(local_grad * (float)packed_dout[k]);
+        }
+
+        store128(dinp + i, packed_dinp);
+    }
+}
+
+// ----------------------------------------------------------------------------
+// kernel launcher
+
+void gelu_backward1(floatX* dinp, const floatX* inp, const floatX* dout, int N, const int block_size) {
+    const int grid_size = ceil_div(N, block_size);
+    gelu_backward1<<<grid_size, block_size>>>(dinp, inp, dout, N);
+    cudaCheck(cudaGetLastError());
+}
+
+void gelu_backward2(floatX* dinp, const floatX* inp, const floatX* dout, int N, const int block_size) {
+    const int grid_size = ceil_div(N, block_size * x128::size);
```

Comment:
maybe add an assert(N % x128::size == 0) here? documents the assumption and we may get a better error message in case we call the kernel wrongly later